component/org.wso2.carbon.identity.oauth2.grant.token.exchange/src/main/java/org/wso2/carbon/identity/oauth2/grant/token/exchange/utils/TokenExchangeUtils.java

```diff
@@ -324,14 +344,41 @@ public static void setAuthorizedUser(OAuthTokenReqMessageContext tokenReqMsgCtx,
                                          IdentityProvider identityProvider, String authenticatedSubjectIdentifier,
                                          JWTClaimsSet claimsSet) throws IdentityOAuth2Exception {
 
-        AuthenticatedUser authenticatedUser;
-        if (Boolean.parseBoolean(IdentityUtil.getProperty(Constants.OAUTH_SPLIT_AUTHZ_USER_3_WAY))) {
-            authenticatedUser = OAuth2Util.getUserFromUserName(authenticatedSubjectIdentifier);
-            authenticatedUser.setAuthenticatedSubjectIdentifier(authenticatedSubjectIdentifier);
-        } else {
-            authenticatedUser =
-                    AuthenticatedUser.createFederateAuthenticatedUserFromSubjectIdentifier(authenticatedSubjectIdentifier);
-            authenticatedUser.setUserName(authenticatedSubjectIdentifier);
+        AuthenticatedUser authenticatedUser = null;
+
+        FederatedAssociationConfig federatedAssociationConfig = identityProvider.getFederatedAssociationConfig();
+        ClaimConfig claimConfig = identityProvider.getClaimConfig();
+
+        ServiceProvider serviceProvider = getServiceProvider(tokenReqMsgCtx);
+        if (serviceProvider.getClaimConfig() != null &&
+                serviceProvider.getClaimConfig().isAlwaysSendMappedLocalSubjectId() &&
+                !Constants.LOCAL_IDP_NAME.equals(identityProvider.getIdentityProviderName())) {
+
+            if (federatedAssociationConfig.isEnabled()) {
+                String subjectIdentifier = resolveSubjectIdentifier(claimsSet,
+                        federatedAssociationConfig.getLookupAttributes());
+                User localUser = getLocalUser(tokenReqMsgCtx, subjectIdentifier);
+                if (localUser != null) {
+                    if (!isUserAssociated(localUser, identityProvider)) {
```

Comment:
IDP subject identifier also should be there, refer comment below at the method signature

Reply:
addressed with ec674ae